### ISSUE TITLE
UpsertDocumentAsync does not throw unique index constraint violation

### ISSUE THREAD

**Issue by mcqueene** [Azure/azure-cosmos-dotnet-v2#642]
**Describe the bug**
UpsertDocumentAsync does not throw unique index constraint violation if the update violates the unique keys.

**To Reproduce**
Create a new container with a unique key. Add a document that meets the unique key constraints. Now update that document with a totally different document content (ex. unique key fields are missing) using the document Id. Verify the new document content violates the keys by trying to add the document with a different Id.  

**Expected behavior**
I would expect that if a document throws "unique index constraint violation" when calling CreateDocumentAsync, it should also throw the same error when calling UpsertDocumentAsync .

**Actual behavior**
While testing changes I altered the structure of the documents I was updating and inadvertently removed the key fields. But using UpsertDocumentAsync never threw an error. I then tried to add a new document using my new structure and the "unique index constraint violation" was thrown.

**Environment summary**
SDK Version: Visual Studio 2017
OS Version (e.g. Windows, Linux, MacOSX) Windows

**Additional context**
Add any other context about the problem here (for example, complete stack traces or logs).

**Comment by Liversage** [comment]
Are you sure that this is not a result of

> Sparse unique keys are not supported. If some unique path values are missing, they are treated as null values, which take part in the uniqueness constraint. Hence, there can only be a single item with null value to satisfy this constraint.

as described in [Unique keys in Azure Cosmos DB](https://docs.microsoft.com/en-us/azure/cosmos-db/unique-keys)?

**Comment by mcqueene** [comment]
> Are you sure that this is not a result of
> 
> > Sparse unique keys are not supported. If some unique path values are missing, they are treated as null values, which take part in the uniqueness constraint. Hence, there can only be a single item with null value to satisfy this constraint.
> 
> as described in [Unique keys in Azure Cosmos DB](https://docs.microsoft.com/en-us/azure/cosmos-db/unique-keys)?

I'm pretty sure this isn't the case. I found this while doing a major rewrite of a custom class using 3 test records with every test run. After days of changes to the class and multiple test runs I had a need to use a different test case. That is when I hit the unique index contraint violation. It took me a while to realize it was because the 4th test case was doing an insert, not an update. If the UpsertDocumentAsync were checking the keys like CreateDocumentAsync then I would have hit this error while running my tests with the first 3 test cases since they all lack the key fields and all have null values for those key fields. It took me an extra long time to figure out the difference because I failed to realize the 4th test case had never been created in the database. If I get time I will try to create a console app to recreate the issue. All that needs to be done is create two simple yet different classes (ex. employee, car) and then create a container with a key that only matches one of the objects (ex. employee_email). Add two valid records to the container (employee), then try and do an upsert on both records using the same Id but with the different class (car). The upsert will work. Then try to create documents using the class that doesn't have the key (car) and it will fail.
